src/libstd/sync/condvar.rs

```diff
@@ -167,6 +163,37 @@ impl Condvar {
         }
     }
 
+    /// Wait on this condition variable for a notification, timing out after a
+    /// specified duration.
+    ///
+    /// The semantics of this function are equivalent to `wait_timeout` except
+    /// that the implementation will repeatedly wait while the duration has not
+    /// passed and the provided function returns `false`.
+    // FIXME this doesn't handle the poison case at all if the user wanted to
+    // unwrap the Err and keep on waiting
+    #[unstable]
+    pub fn wait_timeout_with<'a, T, F>(&self,
+                                       mut guard: MutexGuard<'a, T>,
+                                       dur: Duration,
+                                       mut f: F)
+                                       -> LockResult<(MutexGuard<'a, T>, bool)>
+            where F: FnMut(&mut MutexGuard<'a, T>) -> bool {
+        // This could be made more efficient by pushing the implementation into sys::condvar
+        let start = SteadyTime::now();
+        while !f(&mut guard) {
+            let now = SteadyTime::now();
+            let consumed = &now - &start;
+            let (new_guard, no_timeout) = try!(self.wait_timeout(guard, dur - consumed));
+            guard = new_guard;
+            if !no_timeout {
+                let result = f(&mut guard);
+                return Ok((guard, result));
+            }
+        }
```

Comment:
Instead of duplicating the implementation, could this delegate to `StaticCondvar` below?